RevDiff: Show common/unique files for BASE->selected(B)

gerhardol opened the pull request:
Part of #4154 
Separated from #7899 , the first part of the compare
This is partly to simplify the review, partly to simplify my merges, partly to drive the ideas forward: What more do we want to see?

## Proposed changes

#7720 added the possibility to see differences to BASE-> selected, if the two commits first/selected shared common BASE.

This PR separate the groups for the files that are unique BASE->selected(B) and common for
BASE->selected(B) and BASE->first(A) to give a better diff of branches

#7899 contains the changes to also show the unique files BASE->first(A) and to simplify other structures in GE

## Screenshots

![image](https://user-images.githubusercontent.com/6248932/77946819-3f098880-72c3-11ea-9214-08e2b55eac59.png)

### Before

![image](https://user-images.githubusercontent.com/6248932/77946770-2a2cf500-72c3-11ea-9e91-61ed4a220d82.png)

### After

![image](https://user-images.githubusercontent.com/6248932/77947085-a9bac400-72c3-11ea-9807-1ad002610f1b.png)

See #7899 for BASE->first

## Test methodology 

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -620,91 +621,129 @@ public void SelectStoredNextIndex(int defaultIndex = -1)
+
+                        var revBase = new GitRevision(baseRevGuid);
+                        tuples.Add((revBase,
+                            _diffBaseToB.Text + GetDescriptionForRevision(selectedRev.ObjectId), uniqueBaseToB));
```
Have I missed something why `Unique Diff BASE with first` would not work?
```suggestion
                        tuples.Add((revBase,
                            _diffBaseToB.Text + GetDescriptionForRevision(selectedRev.ObjectId), uniqueBaseToB));
                        tuples.Add((revBase,
                            _diffBaseToB.Text + GetDescriptionForRevision(firstRev.ObjectId), uniqueBaseToA));
```

gerhardol replied:
Unique Diff BASE with first requires that the datastructure changes and affects all forms. (Well the list could be displayed, but no actions would work, as selected/Revision is used for second revisions in diffs it would be confusing.)

As other commits need to be merged prior to #7899, I hope this is a way to speedup reviews...

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -620,91 +621,129 @@ public void SelectStoredNextIndex(int defaultIndex = -1)
             return;
 
-            ObjectId GetRevisionOrHead(ObjectId rev, Lazy<ObjectId> head)
+            ObjectId GetRevisionOrHead(GitRevision rev, Lazy<ObjectId> head)
```
```suggestion
            static ObjectId GetRevisionOrHead(GitRevision rev, Lazy<ObjectId> head)
```

gerhardol replied:
Re# suggests this (I had that in #7899 but reverted it as it is not used elsewhere in GE). Any motivation you know about? Re# is most often correct though.

mstv replied:
I guess `static` was not supported before C#8. IIUIC it indicates that no closure is required, i.e. no variables are bound. It just works on the arguments provided.

gerhardol commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -653,91 +654,129 @@ public void SelectStoredNextIndex(int defaultIndex = -1)
+                        var allBaseToA = Module.GetDiffFilesWithSubmodulesStatus(baseRevGuid, firstRev.ObjectId,
+                            firstRev.FirstParentGuid);
+
+                        var gitItemStatusNameComparer = new GitItemStatusNameComparer();
+                        var commonBaseToAandB = allBaseToB.Intersect(allBaseToA, gitItemStatusNameComparer)
+                            .Except(allAToB, gitItemStatusNameComparer).ToList();
+                        var uniqueBaseToB = allBaseToB.Except(commonBaseToAandB, gitItemStatusNameComparer).ToList();
+                        var uniqueBaseToA = allBaseToA.Except(commonBaseToAandB, gitItemStatusNameComparer).ToList();
```
@RussKie This is the only change of logic in this PR - splitting the BASE diff in unique and common files

RussKie commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -653,91 +654,129 @@ public void SelectStoredNextIndex(int defaultIndex = -1)
+                        // For the following diff:  A->B a,c,d; BASE->B a,b,c; BASE->A a,b,d
+                        // (the file a has unique changes, b has the same change and c,d is changed in one of the branches)
+                        // The following groups will be shown: A->B a,c,d; BASE->B a,c; BASE->A a,d; Common BASE b
+                        var allBaseToB = Module.GetDiffFilesWithSubmodulesStatus(baseRevGuid, selectedRev.ObjectId,
+                            selectedRev.FirstParentGuid);
+                        var allBaseToA = Module.GetDiffFilesWithSubmodulesStatus(baseRevGuid, firstRev.ObjectId,
+                            firstRev.FirstParentGuid);
```
```suggestion
                        var allBaseToB = Module.GetDiffFilesWithSubmodulesStatus(baseRevGuid, selectedRev.ObjectId, selectedRev.FirstParentGuid);
                        var allBaseToA = Module.GetDiffFilesWithSubmodulesStatus(baseRevGuid, firstRev.ObjectId, firstRev.FirstParentGuid);
```

RussKie commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -653,91 +654,129 @@ public void SelectStoredNextIndex(int defaultIndex = -1)
+                    Lazy<ObjectId> head = getRevision != null
+                        ? new Lazy<ObjectId>(() => getRevision(ObjectId.IndexId).FirstParentGuid)
+                        : new Lazy<ObjectId>(() => Module.RevParse("HEAD"));
+                    var baseRevGuid = Module.GetMergeBase(GetRevisionOrHead(firstRev, head),
+                        GetRevisionOrHead(selectedRev, head));
```
```suggestion
                    var baseRevGuid = Module.GetMergeBase(GetRevisionOrHead(firstRev, head), GetRevisionOrHead(selectedRev, head));
```

gerhardol commented:
Adopted the comments (and removed an unused statement), rebased.

> Style related feedback.
Everything perfectly fits on a 1920x screen, wrapping and "ladder" formatting makes it unreadable for me.
Are you using Resharper's vile 80-char wrapping?

Re# is set to 120 chars, I do not believe I have adjusted that.
I thought it was .editor config that wrapped lines when formatting code, so I followed that. No explicit wrapping done.

However, I believe somewhere around 105 chars is a good normal maximum linelength, then it is possible to have several files open at the same time and mobile reviewing is not a pain (and no fiddling with extensions in GitHub web). At work (so not GE code) it is possible to work on the laptop too with 1280x720 or something, the external monitors have at least FHD.

mstv commented:
Do you really want the capitalized "Diff" in e.g. "Combined Diff" and "Unique Diff with"?

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -653,91 +659,119 @@ public void SelectStoredNextIndex(int defaultIndex = -1)
+                    .Skip(1)
+                    .Take(multipleParents)
+                    .Select(parentRev =>
+                        (parentRev, "",  Module.GetDiffFilesWithSubmodulesStatus(parentRev.ObjectId, selectedRev.ObjectId, selectedRev.FirstParentGuid))));
```
micro nit:
```suggestion
                        (parentRev, "", Module.GetDiffFilesWithSubmodulesStatus(parentRev.ObjectId, selectedRev.ObjectId, selectedRev.FirstParentGuid))));
```

gerhardol replied:
> It's getting less to compare.

fixing, squashing after a few others are merged tomorrow. Too many rebases with all related changes lately, but now most are being merged, except the two big PRs
It would have been easier if I had these changes when starting to implement them...

gerhardol commented:
> Do you really want the capitalized "Diff" in e.g. "Combined Diff" and "Unique Diff with"?

"Combined Diff" was like this before, I followed that for the added strings, but that is not following the GE standard. Will change (also "Combined Diff").

gerhardol commented on GitCommands/Git/GitItemStatusNameComparer.cs:
```diff
@@ -0,0 +1,17 @@
+
+        public int GetHashCode(GitItemStatus obj)
+        {
+            return obj.Name.GetHashCode();
```
Set to 0 if Name is null

RussKie commented on GitCommands/Git/GitItemStatusNameComparer.cs:
```diff
@@ -0,0 +1,17 @@
+
+        public int GetHashCode(GitItemStatus obj)
+        {
+            return obj.Name?.GetHashCode() ?? 0;
```
`obj` can too be `null`
```suggestion
            return obj?.Name?.GetHashCode() ?? 0;
```

gerhardol replied:
Not what I can see from the docs or in any example, but changing

gerhardol commented:
Added some review comments from #7899 to this PR
Also changed the existing IComparer usage (IComparable interface) and IEqualityComparer.
Previously, the default Comparer compared on Name/OldName only which is a little misleading so I moved that to a separate file (similar to GitItemStatusFileExtensionComparer).
For this PR it does not matter, for another I wanted the IEqualityComparer to operate on the internal flags too so I needed to move that to GitItemStatus.
GitItemStatusNameComparer therefore got a different implementation
All changes to be squashed

@RussKie I would like to include this in 4.0

gerhardol commented:
More or less reverted the EqualityComparer as that did not work when comparing WorkTree files (Staged status can not be evaluated).
Forced pushed the changes, if you are interested in the changes see d122a7536537f98a4a294f76d718485a1f9b086e

gerhardol commented:
@msftbot merge in 24 hours
